### DISCUSSION TITLE
Multiple Library Codegens?

### DISCUSSION THREAD

**Discussion by iBotPeaches** [reactwg/react-native-new-architecture#158] [Q&A] [answered]
I engulfed a private library that was working an upgrade to New Arch - it has never been deployed/merged. The intending application is now 73.

So I took it for a spin with `pod install` and it crashed out quick with:

```
[Codegen] Done.
TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received undefined
    at new NodeError (node:internal/errors:405:5)
    at validateString (node:internal/validators:162:11)
    at Object.join (node:path:1171:7)
    at node_modules/react-native/scripts/codegen/generate-artifacts-executor.js:398:40
```

Which is probably because I have an array of libraries vs one I'm guessing:

```json
  "codegenConfig": [
    {
      "name": "RNBluetoothSerialSpec",
      "type": "modules",
      "jsSrcsDir": "src",
      "android": {
        "javaPackageName": "com.xxx.xxx"
      }
    },
    {
      "name": "RNAcessoryManagerSpec",
      "type": "modules",
      "jsSrcsDir": "src"
    }
  ],
```

Though I'm looking at the original ticket where this development started and it has quoted this is a valid example:

```json
"codegenConfig": {
        "libraries": [
          {
            "name": "libName1",
            "type": "all|components|modules",
            "jsSrcsRoot": "libName1/js"
          },
          {
            "name": "libName2",
            "type": "all|components|modules",
            "jsSrcsRoot": "libName2/src"
          }
        ]
      }
```

Ultimately I struggle in docs to find if this is a supported behavior now, so looking for some help. My guess is I might have to unify these libraries to 1 between iOS/Android - However, this is Bluetooth so between External Accessory on iOS and `android.bluetooth.*` on Android I don't see an elegant way to do that without some no-op methods on both platforms.

A bit lost as I can't find anything in the [Codegen docs](https://reactnative.dev/docs/the-new-architecture/pillars-codegen) of the schema itself of this property.

Someone able to just set me straight?

**Comment by vonovak** [comment, marked as answer]
I believe you can specify one entry for `codegenConfig` in package.json

```
  "codegenConfig": {
    "name": "AbcCGen",
    "type": "all",
    "jsSrcsDir": "src/spec",
    "android": {
      "javaPackageName": "com.abc"
    }
  },
```

and then put multiple specs into `src/spec`. Codegen will pick it up.

**Reply by cortinico** [reply]
That's the correct answer. We used to support an array of dependencies inside `codegenConfig` but we moved away from it as it was divergent between Android & iOS and was complicating the codegen code

**Reply by lovegaoshi** [reply]
thx! I think all spec files need to start with `Native`. was struggling with specs not showing up until I renamed the added spec file to NativeXXXModule (I dont think module here matters), then it worked.

**Reply by NikhilRW** [reply]
Hi I Have A Question Can We Put C++ Module And Java Based Module In Same Codegen Config And Build The App ?
